R/plotDists.R

```diff
@@ -25,101 +25,161 @@
+#'
+#' plot{...} functions are generated programmatically so their \code{body} will not
+#' be clear unless you inspect that environment directly
+#'
+#' @examples
+#' body(plotLogNormal)
+#' environment(plotLogNormal)$funCall
 NULL
 
+dpareto <- function(x, xm, alpha) ifelse(x > xm , alpha * xm ** alpha / (x ** (alpha + 1)), 0)
+ppareto <- function(q, xm, alpha) ifelse(q > xm , 1 - (xm / q) ** alpha, 0 )
+qpareto <- function(p, xm, alpha) ifelse(p < 0 | p > 1, NaN, xm * (1 - p) ** (-1 / alpha))
+rpareto <- function(n, xm, alpha) qpareto(runif(n), xm, alpha)
+
+qinvgamma_ <- function(area, shape, scale) {
+  if((1 - area) <= .Machine$double.eps) {
+    return(Inf)
+  }
+  if(area <= .Machine$double.eps) {
+    return(0)
+  }
+  return(1 / qgamma(1 - area, shape, scale))
+}
+qinvgamma <- Vectorize(qinvgamma_, vectorize.args = 'area')
+
+dinvgamma_ <- function(x, shape, scale) {
+    if (shape <= 0 | scale <= 0) {
+      stop("Shape or scale parameter negative in dinvgamma().\n")
+    }
+    if(x == 0) return(0)
+    alpha <- shape
+    beta <- scale
+    log.density <- alpha * log(beta) - lgamma(alpha) - (alpha + 1) * log(x) - (beta / x)
+    return(exp(log.density))
+}
+dinvgamma <- Vectorize(dinvgamma_, vectorize.args = 'x')
+
+plotDist_ <- function(support, hseq, dist, params) {
+
+  discretes <- c('Poisson')
+
+  ribbon_or_bar <- ggplot2::geom_ribbon(ymin = 0,
+                                        ymax = hseq,
+                                        size = 2,
+                                        color = I("lightblue"),
+                                        fill = "lightgreen",
+                                        alpha = .25)
+
+  if(dist %in% discretes) ribbon_or_bar <- ggplot2::geom_bar(stat = "identity",
+                                                             color = I("lightblue"),
+                                                             fill = "lightgreen",
+                                                             alpha = .25,
+                                                             size = 2)
+
+  paramList <- sapply(names(params), function(p) paste(p, params[p], sep = " = ", collapse = ""), USE.NAMES = FALSE)
+  paramList <- paste0(paramList, collapse = ", ")
+
+  p <- ggplot2::qplot(x = support, y = hseq, geom = "line") +
+    ggplot2::xlab(NULL) +
+    ggplot2::ylab('PDF') +
+    ggplot2::ggtitle(paste0(
+      dist,
+      ' Probability Density Function for Parameters: ',
+      paramList)) +
+    ribbon_or_bar +
+    theme_bayesAB()
+
+  p
+
+}
+
+alist2 <- function(args) {
+  res <- replicate(length(args), substitute())
+  names(res) <- args
+  res
+}
+
+plotDist <- function(dist, name, args) {
+  makeDistFunc <- function(t) eval(parse(text = paste0(t, dist)))
+  qDist <- makeDistFunc('q')
+  dDist <- makeDistFunc('d')
+
+  args <- sapply(args, as.name, USE.NAMES = FALSE)
+  funCall <- list(list(c(0, 0.01, .99, 1)), args)
```

Comment:
I didn't really see the point of opening these up to the end user since this is just checking whether the `0` and `1` probabilities have finite bounds for the PDF, and using `0.01` and `0.99` if not. I could see an argument for moving this out of the inner function though. The problem with closures that have dynamically assigned body/formals is that choosing what values stay in the parent environment vs the closure seem a bit arbitrary.